Lib/zipfile.py

```diff
@@ -1729,6 +1740,17 @@ def _extract_member(self, member, targetpath, pwd):
              open(targetpath, "wb") as target:
             shutil.copyfileobj(source, target)
 
+        if preserve_permissions == PreserveMode.NONE:
+            return targetpath
+
+        if preserve_permissions == PreserveMode.SAFE:
+            mode = (member.external_attr >> 16) & 0o777
+
+        if preserve_permissions == PreserveMode.ALL:
+            mode = (member.external_attr >> 16) & 0xFFFF
+
+        os.chmod(targetpath, mode)
```

Comment:
`member.external_attr` only contains info about the UNIX permissions when the ZIP file was created on unix-like system. If the ZIP file was created on Windows, then the value of `member.external_attr >> 16` will be `0`, so it is bad because you end doing `os.chmod(targetpath,0)` here, effectively setting all the permissions denied. You can check that with a ZIP file that was created on Windows ([example here](https://communities.vmware.com/wbsdv95928/attachments/wbsdv95928/3026/388/1/Adobe%20Flash%20Player.zip))

You have to check that `ZipInfo.create_system != 0` before trying to access `member.external_attr >> 16`

Also, when running on Windows it doesn't make much sense to do `os.chmod(targetpath, mode)` for the UNIX attributes. Those can't be applied there. For Windows you have to either avoid doing the `os.chmod()` call or at least call it with the DOS FAT attributes from `member.external_attr` which can be obtained by getting the lowest byte of it: `member.external_attr & 0xFF`

If you want more info about the `external_attr` data see the comment on [this WPT PR](https://github.com/web-platform-tests/wpt/pull/33666/files)

Reply:
a nitpick: doing `mode = (member.external_attr >> 16) & 0xFFFF` is not needed. It is enough doing `mode = member.external_attr >> 16` because the `external_attr` data has a maximum size of 4 bytes.

Reply:
Good call, I'll get around to doing those later today.

Reply:
I think running `os.chmod` on Windows might still be ok. From the documentation for `os`, it looks like windows some flags which chmod will be setting here.

https://github.com/python/cpython/blob/48ec61a89a959071206549819448405c2cea61b0/Doc/library/os.rst#L1838-L1842

https://docs.python.org/3/library/os.html#os.chmod

Reply:
> You have to check that ZipInfo.create_system != 0 before trying to access member.external_attr >> 16

I think a potential problem is that zip files that were created on Windows systems with permissions set will have their ignored. But right now, I still think your suggestion might be the best way to deal with this.

As a patch, how is this?

```patch
diff --git a/Lib/zipfile.py b/Lib/zipfile.py
index 6b955dbdfe..f13531da9b 100644
--- a/Lib/zipfile.py
+++ b/Lib/zipfile.py
@@ -1740,7 +1740,7 @@ def _extract_member(self, member, targetpath, pwd,
              open(targetpath, "wb") as target:
             shutil.copyfileobj(source, target)
 
-        if preserve_permissions == PreserveMode.NONE:
+        if zinfo.create_system == 0 or preserve_permissions == PreserveMode.NONE:
             return targetpath
 
         if preserve_permissions == PreserveMode.SAFE:
```

Reply:
`os.chmod` works on Windows, but you can't pass the UNIX values (`member.external_attr >> 16`) to that call in the `mode` variable, as those values are UNIX-specific and Windows won't understand them. On Windows permissions don't work like in a UNIX-like system. There is ok to not set any specific permissions since executing a program is just a matter of having a `.exe` suffix and not a matter of permissions. Also files are readable/writable by default.

So my suggestion is to avoid calling ``os.chmod`` at all when running on Windows. Suggested patch:

```diff
diff --git a/Lib/zipfile.py b/Lib/zipfile.py
index 6b955dbdfe..f13531da9b 100644
--- a/Lib/zipfile.py
+++ b/Lib/zipfile.py
@@ -1740,7 +1740,7 @@ def _extract_member(self, member, targetpath, pwd,
              open(targetpath, "wb") as target:
             shutil.copyfileobj(source, target)
 
-        if preserve_permissions == PreserveMode.NONE:
+        if member.create_system == 0 or sys.platform == 'win32' or preserve_permissions == PreserveMode.NONE:
             return targetpath
 
         if preserve_permissions == PreserveMode.SAFE:
```

It would be a good idea to also include some notes on the documentation saying that permissions only apply to UNIX-like systems when un-compressing a file that was created on a UNIX-like system.

Reply:
After re-reading your comment: if calling `os.chmod()` on windows with the UNIX values (`member.external_attr >> 16`) ignores all the values from UNIX that it doesn't understand and only applies the read-only byte then your suggested patch is better than my suggestion. It makes sense to mark the file as read-only on Windows if the ZIP file was crated on a UNIX-like system and the UNIX permissions indicates it should be read-only.